internal/cluster/discovery.go

```diff
@@ -0,0 +1,179 @@
+	}
+}
+
+func (d *DiscoveryStrategyAdapter) Addresses(ctx context.Context) ([]pubcluster.Address, error) {
+	d.lg.Trace(func() string {
+		return "cluster.DiscoveryStrategyAdapter.Addresses"
+	})
+	if err := d.ensureStarted(ctx); err != nil {
+		return nil, err
+	}
+	nodes, err := d.strategy.DiscoverNodes(ctx)
+	if err != nil {
+		return nil, err
+	}
+	res := make([]pubcluster.Address, 0, len(nodes))
+	for _, node := range nodes {
+		if node.PrivateAddr == "" && node.PublicAddr == "" {
+			continue
+		}
+		if node.PrivateAddr == "" {
+			res = append(res, (pubcluster.Address)(node.PublicAddr))
+			continue
+		}
+		if node.PublicAddr == "" {
+			res = append(res, (pubcluster.Address)(node.PrivateAddr))
+			continue
+		}
+		if d.usePublicIP {
+			res = append(res, (pubcluster.Address)(node.PublicAddr))
+			continue
+		}
+		res = append(res, (pubcluster.Address)(node.PrivateAddr))
+	}
+	return res, nil
+}
+
+func (d *DiscoveryStrategyAdapter) TranslateMember(ctx context.Context, member *pubcluster.MemberInfo) (addr pubcluster.Address, err error) {
+	d.lg.Trace(func() string {
+		return fmt.Sprintf("cluster.DiscoveryStrategyAdapter.TranslateMember: %s %s", member.UUID.String(), member.Address)
+	})
+	if err = d.ensureStarted(ctx); err != nil {
+		return
+	}
+	addr, ok := d.translate(member.Address)
+	if !ok {
+		nodes, err := d.strategy.DiscoverNodes(ctx)
+		if err != nil {
+			return addr, err
+		}
+		d.updateTranslation(nodes)
+		addr, ok = d.translate(member.Address)
+	}
+	if !ok {
+		return addr, hzerrors.ErrAddressNotFound
+	}
+	return addr, nil
+}
+
+func (d *DiscoveryStrategyAdapter) ensureStarted(ctx context.Context) error {
+	if atomic.LoadInt32(&d.started) != discoveryStrategyAdapterStarted {
+		// discovery strategy was not started yet, try to start it
+		if err := d.start(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *DiscoveryStrategyAdapter) start(ctx context.Context) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	if atomic.LoadInt32(&d.started) == discoveryStrategyAdapterStarted {
+		// strategy was already started
+		return nil
+	}
+	// run strategy's Start method if it implements that.
+	if starter, ok := d.strategy.(discovery.StrategyStarter); ok {
+		d.lg.Debug(func() string {
+			return "Starting discovery strategy"
+		})
+		if err := starter.Start(ctx, d.opts); err != nil {
+			return err
+		}
+	}
+	atomic.StoreInt32(&d.started, discoveryStrategyAdapterStarted)
+	return nil
+}
+
+func (d *DiscoveryStrategyAdapter) translate(addr pubcluster.Address) (pubcluster.Address, bool) {
+	d.mu.RLock()
+	addr, ok := d.translator[addr]
+	d.mu.RUnlock()
+	return addr, ok
+}
+
+func (d *DiscoveryStrategyAdapter) updateTranslation(nodes []discovery.Node) {
+	translator := make(map[pubcluster.Address]pubcluster.Address, len(nodes))
+	for _, node := range nodes {
+		// map private addresses to public, if usePublicIP
+		// map public IPs to public
+		if node.PrivateAddr == "" && node.PublicAddr == "" {
+			continue
+		}
```

Comment:
Do we have these checks in Java side? I think these are confusing. How these can be empty? Here is the code to get privateToPublicAddress map in Java (there is no conditionals): https://github.com/hazelcast/hazelcast/blob/7c3894b1474dba56648cf10ad8441615486fcdea/hazelcast/src/main/java/com/hazelcast/client/impl/clientside/ClusterDiscoveryServiceBuilder.java#L161

Reply:
The discovery strategy can send empty public/private addresses. Instead of panicking we are ignoring them.

Reply:
It is an option as well but different from Java. Are you sure we want a different behaviour?

Reply:
As far as I see in the Java client there's that comment about private address cannot be null, but there's no check for that. So a discovery strategy may return a null private address. The difference in the Go client is, it checks this and does not update the translation table with the null address.